### PR TITLE
Feat/enable multiple custom cost files

### PR DESCRIPTION
**Pull request by fwiedemair**
closes https://github.com/AGGM-AG/pypsa-at-planning/issues/244

## Changes proposed in this Pull Request
Problem:
The costs/custom_cost_fn in PyPSA-AT differs from the one in PyPSA-DE. However we want the information from both files for our cost.

Implementation:
A new config option has been added to pass a list of cost_files. A new rule+script has been added to combine the listed cost files into a common resource file to be used for cost-processing rules.

## Checklist

**Required:**
- [x] Changes are tested locally and behave as expected.
- [x] Code and workflow changes are documented.
- [x] A brief description of the changes has been added to `Changelog.md`.
- [x] All Sourcery Bot review suggestions have been implemented or discarded with an explanation.
- [x] All github actions succeed.

**If applicable:**
- [ ] Changes in configuration options are reflected in `scripts/lib/validation`.
- [ ] For new data sources or versions, [these instructions](https://pypsa-eur.readthedocs.io/en/latest/data_sources.html) have been followed.
- [ ] New rules are documented in the appropriate `docs-at/` files.

## Summary by Sourcery

Add support for combining multiple custom cost CSV files into a single resource and wire it into the PyPSA-AT electricity build workflow.

New Features:
- Allow configuring a list of custom cost files that are merged into a unified custom_cost_fn resource used in cost processing.

Enhancements:
- Introduce a PyPSA-AT-specific build rule and script to generate a combined custom cost file, with configuration validation updates and a dedicated electricity build override.
- Document the new custom cost file list configuration in the AT changelog.

Tests:
- Add unit tests covering multi-file combination behavior, duplicate handling, column preservation, and error conditions for the custom cost file builder.

### CODE REVIEW

File: scripts/lib/validation/config/costs.py
Comment (pworschischek-aggmag) [inline scripts/lib/validation/config/costs.py:1]: Not sure if we need to extend the upstream file if on `config.at.yaml` is extended. AFAIK the `lib/validation` only applies to `config.default.yaml`

File: config/config.at.yaml
Comment (pworschischek-aggmag) [inline config/config.at.yaml:1]: we could simply access the costs list feature if additional cost files are listed under `costs.custom_cost_fn_list` and skip if the list is empty. Or simply pass a list to  `costs.custom_cost` directly and check types in the snakemake `branch()` helper

File: rules/pypsa-at/build.smk [hunk 1]
```diff
@@ -7,6 +7,28 @@ Build rules for AT-specific datasets.
 """
 
 
+rule build_custom_cost_fn:
+    input:
+        custom_cost_files=branch(
+            config_provider("costs", "use_list"),
+            config_provider("costs", "custom_cost_fn_list"),
```
Comment (pworschischek-aggmag) [inline rules/pypsa-at/build.smk:14]: I guess we need to make our costs file scenario dependent too

File: rules/pypsa-at/build.smk [hunk 2]
```diff
@@ -7,6 +7,28 @@ Build rules for AT-specific datasets.
...
+            lambda w: [config_provider("costs", "custom_cost_fn")(w)],
+        ),
+    output:
+        custom_cost_fn=resources("custom_cost_fn.csv"),
```
Comment (pworschischek-aggmag) [inline rules/pypsa-at/build.smk]: Lets refactor the output file name `custom_cost_fn.csv` to something more descriptive. maybe `custom_cost_at.csv` is sufficient. Lets discuss

File: test/test_build_custom_cost_fn.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+"""Unit tests for the custom cost file builder script."""
+
+from importlib import import_module
+
+import pandas as pd
+import pytest
+
+from scripts._helpers import mock_snakemake
+
+main = import_module("scripts.pypsa-at.build_custom_cost_fn").main
```
Comment (pworschischek-aggmag) [inline test/test_build_custom_cost_fn.py]: concise workaround. I like it

File: test/test_build_custom_cost_at.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import pytest
+
+from scripts._helpers import mock_snakemake
+
+main = import_module("scripts.pypsa-at.build_custom_cost_fn").main
+
+
+def test_combine_two_files_with_overlapping_entries_keeps_first(tmp_path):
+    """
+    Test that combining two CSV files retains the first file's values for
```
Comment (pworschischek-aggmag) [inline test/test_build_custom_cost_at.py:19]: Should add a note in the config that entries from the first file in the list persist and subsequent entries will be dropped.

**Review by pworschischek-aggmag — CHANGES_REQUESTED** [review]
No bigger issues. Just some remarks on output file name and some on how to enter the feature in snakemake

File: scripts/lib/validation/config/costs.py
Comment (fwiedemair) [reply scripts/lib/validation/config/costs.py:1]: Removed

File: rules/pypsa-at/build.smk (hunk 2 above)
Comment (fwiedemair) [reply rules/pypsa-at/build.smk]: Renamed

File: test/test_build_custom_cost_fn.py (hunk 3 above)
Comment (fwiedemair) [reply test/test_build_custom_cost_fn.py]: Thanks!

File: config/config.at.yaml
Comment (fwiedemair) [reply config/config.at.yaml:1]: For now stays as it

File: rules/pypsa-at/build.smk (hunk 1 above)
Comment (fwiedemair) [reply rules/pypsa-at/build.smk:14]: Moved to scenario folder

File: test/test_build_custom_cost_at.py (hunk 4 above)
Comment (fwiedemair) [reply test/test_build_custom_cost_at.py:19]: Added

**Review by pworschischek-aggmag — APPROVED** [review]
Ready to merge
